src/adapter/src/catalog/state.rs

```diff
@@ -2383,6 +2383,55 @@ impl CatalogState {
     pub fn storage_metadata(&self) -> &StorageMetadata {
         &self.storage_metadata
     }
+
+    /// For the Sources ids in `ids`, return the compaction windows for all `ids` and additional ids
+    /// that propagate from them. Specifically, `ids` contains a source, all of its subsources will
+    /// be added to the result.
+    ///
+    /// # Panics
+    /// Panics if any of `ids` is not a `Source` item type.
+    pub fn source_compaction_windows(
+        &self,
+        ids: impl IntoIterator<Item = GlobalId>,
+    ) -> BTreeMap<CompactionWindow, BTreeSet<GlobalId>> {
+        let mut cws: BTreeMap<CompactionWindow, BTreeSet<GlobalId>> = BTreeMap::new();
+        let mut ids = VecDeque::from_iter(ids);
+        let mut seen = BTreeSet::new();
+        while let Some(id) = ids.pop_front() {
+            if !seen.insert(id) {
+                continue;
+            }
+            let entry = self.get_entry(&id);
+            let source = entry.source().expect("must be source");
+            let source_cw = source.custom_logical_compaction_window.unwrap_or_default();
+            match source.data_source {
+                DataSourceDesc::Ingestion(_) => {
+                    // For sources, look up each dependent subsource and propagate.
+                    cws.entry(source_cw).or_default().insert(id);
+                    ids.extend(entry.used_by());
```

Comment:
naively I thought `used_by()` would return Views that reference the source, but maybe not?

Reply:
Good find! Added a test for this that panic'd. Fixed.